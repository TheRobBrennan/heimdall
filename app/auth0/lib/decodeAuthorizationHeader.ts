import { decodeAuth0JWT } from "./decodeAuth0JWT"

export const decodeAuthorizationHeader = async (
  authorizationHeaderValue: string
) => {
  try {
    // Use array destructuring to disregard the first item. We only care about the JWT.
    const [_, token] = authorizationHeaderValue.split("Bearer ")

    // DEBUG
    console.log(`
    [DEBUG] Auth0 decodeAuthorizationHeader received authorization header value:

    ${authorizationHeaderValue}
    `)

    // Decode and verify JWT using Auth0 keys
    const decodedAuth0JWT = await decodeAuth0JWT(token)
    console.log(
      `Decoded Auth0 JWT: ${JSON.stringify(decodedAuth0JWT, null, 2)}`
    )

    // Return the decoded JWT
    return decodedAuth0JWT
  } catch (e) {
    return console.error(
      `Unable to decode authorizationHeaderValue: ${authorizationHeaderValue}`
    )
  }
}
