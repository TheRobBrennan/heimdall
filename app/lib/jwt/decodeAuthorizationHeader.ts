import { decodeJWT } from "./decodeJWT"
import { decodeAuth0JWT } from "../../auth0/lib"

export const decodeAuthorizationHeader = async (authorizationHeader) => {
  try {
    // Use array destructuring to disregard the first item. We only care about the JWT.
    const [_, token] = authorizationHeader.split("Bearer ")

    // This decodes ANY well-formed JWT and does not vouch for validity or authenticity
    const decoded = decodeJWT(token)
    console.log(
      `Decoded generic JWT without token verification: ${JSON.stringify(
        decoded,
        null,
        2
      )}`
    )

    // Decode and verify JWT using Auth0 keys
    const decodedAuth0JWT = await decodeAuth0JWT(token)
    console.log(
      `Decoded Auth0 JWT: ${JSON.stringify(decodedAuth0JWT, null, 2)}`
    )
  } catch (e) {
    return console.error(
      `Unable to decode authorizationHeader: ${authorizationHeader}`
    )
  }
}
