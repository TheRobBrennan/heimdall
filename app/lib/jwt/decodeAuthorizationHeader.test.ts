import { cleanup } from "@testing-library/react"

import { A_WELL_FORMED_BUT_UNTRUSTED_AND_UNVERIFIED_JWT_TOKEN } from "./index"
import { decodeAuthorizationHeader as func } from "./decodeAuthorizationHeader"

describe("Our generic JWT decodeAuthorizationHeader lib", () => {
  afterEach(cleanup)
  describe('when supplied with a valid "Bearer <token>" string', () => {
    it("should receive a decoded JWT object", async () => {
      const mockHttpAuthorizationHeaderValue = `Bearer ${A_WELL_FORMED_BUT_UNTRUSTED_AND_UNVERIFIED_JWT_TOKEN}`
      const result = await func(mockHttpAuthorizationHeaderValue)
      expect(result).toMatchSnapshot()
    })
  })
  describe('when supplied with a string that does not contain a valid "Bearer <token>"', () => {
    it("should return undefined", async () => {
      const mockHttpAuthorizationHeaderValue = `Bearer fakeToken`
      const result = await func(mockHttpAuthorizationHeaderValue)
      expect(result).toEqual(undefined)
    })
  })
  describe("when supplied with an unexpected string", () => {
    it("should return undefined", async () => {
      const mockHttpAuthorizationHeaderValue = `Here is an example string that is neither a JWT token nor a valid string in the HTTP Authorization header`
      const result = await func(mockHttpAuthorizationHeaderValue)
      expect(result).toEqual(undefined)
    })
  })
})
