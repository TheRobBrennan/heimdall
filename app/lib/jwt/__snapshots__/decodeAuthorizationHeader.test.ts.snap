// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Our generic JWT decodeAuthorizationHeader lib when supplied with a valid "Bearer <token>" string should receive a decoded JWT object 1`] = `undefined`;
